x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/json_diff/mark_edits.tsx

```diff
@@ -0,0 +1,214 @@
+  diffCss: StringDiffFn;
+}
+
+const jsDiff: JsDiff = diff;
+
+export enum DiffMethod {
+  CHARS = 'diffChars',
+  WORDS = 'diffWords',
+  WORDS_WITH_SPACE = 'diffWordsWithSpace',
+  LINES = 'diffLines',
+  TRIMMED_LINES = 'diffTrimmedLines',
+  SENTENCES = 'diffSentences',
+  CSS = 'diffCss',
+}
+
+function findChangeBlocks(changes: ChangeData[]): ChangeData[][] {
+  const start = findIndex(changes, (change) => !isNormal(change));
+
+  if (start === -1) {
+    return [];
+  }
+
+  const end = findIndex(changes, (change) => !!isNormal(change), start);
+
+  if (end === -1) {
+    return [changes.slice(start)];
+  }
+
+  return [changes.slice(start, end), ...findChangeBlocks(changes.slice(end))];
+}
+
+function groupDiffs(diffs: Diff[]): [Diff[], Diff[]] {
+  return diffs.reduce<[Diff[], Diff[]]>(
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    ([oldDiffs, newDiffs], diff) => {
+      const [type] = diff;
+
+      switch (type) {
+        case DmpChangeType.INSERT:
+          newDiffs.push(diff);
+          break;
+        case DmpChangeType.DELETE:
+          oldDiffs.push(diff);
+          break;
+        default:
+          oldDiffs.push(diff);
+          newDiffs.push(diff);
+          break;
+      }
+
+      return [oldDiffs, newDiffs];
+    },
+    [[], []]
+  );
+}
+
+function splitDiffToLines(diffs: Diff[]): Diff[][] {
+  return diffs.reduce<Diff[][]>(
+    (lines, [type, value]) => {
+      const currentLines = value.split('\n');
+
+      const [currentLineRemaining, ...nextLines] = currentLines.map(
+        (line: string): Diff => [type, line]
+      );
+      const next: Diff[][] = [
+        ...lines.slice(0, -1),
+        [...lines[lines.length - 1], currentLineRemaining],
+        ...nextLines.map((line) => [line]),
+      ];
+      return next;
+    },
+    [[]]
+  );
+}
+
+function diffsToEdits(diffs: Diff[], lineNumber: number): RangeTokenNode[] {
+  const output = diffs.reduce<[RangeTokenNode[], number]>(
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    (output, diff) => {
+      const [edits, start] = output;
+      const [type, value] = diff;
+      if (type !== DmpChangeType.EQUAL) {
+        const edit: RangeTokenNode = {
+          type: 'edit',
+          lineNumber,
+          start,
+          length: value.length,
+        };
+        edits.push(edit);
+      }
+
+      return [edits, start + value.length];
+    },
+    [[], 0]
+  );
+
+  return output[0];
+}
+
+function convertToLinesOfEdits(linesOfDiffs: Diff[][], startLineNumber: number) {
+  return flatMap(linesOfDiffs, (diffs, i) => diffsToEdits(diffs, startLineNumber + i));
+}
+
+function diffBy(diffMethod: DiffMethod, x: string, y: string): [Diff[], Diff[]] {
+  const jsDiffChanges: Change[] = jsDiff[diffMethod](x, y);
+  const diffs: Diff[] = diff.convertChangesToDMP(jsDiffChanges);
+
+  if (diffs.length <= 1) {
+    return [[], []];
+  }
+
+  return groupDiffs(diffs);
+}
+
+function diffChangeBlock(
```

Comment:
```suggestion
/**
 * @param {ChangeData[]} changes - An array of ChangeData representing the changes in the block.
 * @param {DiffMethod} diffMethod - The method used for performing the diff.
 * @returns {[RangeTokenNode[], RangeTokenNode[]]} An array containing two arrays of RangeTokenNodes representing the old and new source of the change block.
 *
 * @description
 * Converts the given ChangeData array to two strings representing the old source and new source of a change block. 
 * The format of the strings is as follows:
 *
 * Old Source:
 * "created_at": "2023-11-20T16:47:52.801Z",
 * "created_by": "elastic",
 * ...
 *
 * New Source:
 * "created_at": "1970-01-01T00:00:00.000Z",
 * "created_by": "",
 * ...
 */
function diffChangeBlock(
  changes: ChangeData[],
  diffMethod: DiffMethod
): [RangeTokenNode[], RangeTokenNode[]] {
```

Reply:
Just suggesting to move your explanatory comment into a JSDoc

Reply:
Done